File: python/examples/core/02-auth-flows/main.py
"""Example 02: three Canvus authentication flows side-by-side.

Runs each path in sequence and reports per-path success/failure. A failure in
one path does not abort the next, so a partial setup (e.g. API key present
but login credentials absent) still produces useful output.

Flows:
    1. **API key** — `Private-Token` header (long-lived).
    2. **Email + password login** — `POST /users/login`, returns a session
       token. Sends ``email`` ONLY; sending ``username`` causes the live
       server to reject the request (see VERIFIED-CORRECTIONS.md §4).
    3. **Programmatic access-token lifecycle** — create / list / delete a
       per-user token under the user authenticated by flow 1.
"""

from __future__ import annotations

import asyncio
import sys
from datetime import UTC, datetime

import structlog
from pydantic import Field, ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict

from canvus_sdk import (
    APIError,
    AuthError,
    CanvusError,
    Client,
    configure_logging,
)

logger = structlog.get_logger(__name__)


class AuthSettings(BaseSettings):
    """Environment-backed configuration for the auth-flows example."""

    model_config = SettingsConfigDict(
        env_prefix="CANVUS_",
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore",
    )

    api_url: str = Field(..., description="Base Canvus URL.")
    api_key: str | None = Field(default=None, description="API key for flow 1 and flow 3.")
    email: str | None = Field(default=None, description="Login email for flow 2.")
    password: str | None = Field(default=None, description="Login password for flow 2.")


async def run_api_key(settings: AuthSettings) -> bool:
    """Flow 1: authenticate with an API key.

    Returns True on success, False otherwise. We can't introspect the
    authenticated user without /users/me (which v1.2 doesn't expose), so
    a successful canvas list is the strongest proof we can produce here.
    """
    if not settings.api_key:
        logger.warning("flow_api_key skipped", reason="CANVUS_API_KEY not set")
        return False
    async with Client(settings.api_url, settings.api_key) as client:
        try:
            canvases = await client.canvases.list()
        except AuthError as e:
            logger.error("flow_api_key failed", status_code=e.status_code)
            return False
        except APIError as e:
            logger.error("flow_api_key api error", status_code=e.status_code)
            return False
        logger.info("flow_api_key ok", canvas_count=len(canvases))
    return True


async def run_login(settings: AuthSettings) -> tuple[str, str] | None:
    """Flow 2: log in with email + password.

    The server rejects requests that double-key email and username; we send
    only ``email`` per VERIFIED-CORRECTIONS.md §4.
    """
    if not (settings.email and settings.password):
        logger.warning(
            "flow_login skipped",
            reason="CANVUS_EMAIL or CANVUS_PASSWORD missing",
        )
        return None
    # The Client needs *some* api key to construct; we'll use an empty one
    # because /users/login does not require auth, then swap in the issued
    # token if we want to make a follow-up call.
    async with Client(settings.api_url, api_key="bootstrap") as client:
        try:
            login = await client.auth.login(
                email=settings.email,
                password=settings.password,
            )
        except AuthError as e:
            logger.error("flow_login failed", status_code=e.status_code)
            return None
        except APIError as e:
            logger.error("flow_login api error", status_code=e.status_code)
            return None
    user_id = login.user.id if login.user else None
    logger.info("flow_login ok", user_id=user_id, has_token=bool(login.token))
    if user_id is None or not login.token:
        return None
    return (str(user_id), login.token)


async def run_token_lifecycle(
    settings: AuthSettings,
    user_id: str,
    auth_token: str,
) -> None:
    """Flow 3: create + list + delete a programmatic access token.

    Uses the session token returned by flow 2's login as the auth credential.
    Caller passes ``user_id`` (also from login) explicitly because v1.2
    does not expose ``/users/me``.
    """
    token_name = f"example-02 {datetime.now(UTC).isoformat()}"
    async with Client(settings.api_url, auth_token) as client:
        try:
            created = await client.auth.create_token(user_id, name=token_name)
        except APIError as e:
            logger.error(
                "flow_token_lifecycle create failed",
                status_code=e.status_code,
            )
            return
        logger.info(
            "flow_token_lifecycle created",
            token_id=created.id,
            plain_token_prefix=(created.plain_token or "")[:6] + "…",
        )

        try:
            tokens = await client.auth.list_tokens(user_id)
            logger.info("flow_token_lifecycle listed", count=len(tokens))
        except APIError as e:
            logger.error("flow_token_lifecycle list failed", status_code=e.status_code)

        try:
            await client.auth.delete_token(user_id, created.id)
            logger.info("flow_token_lifecycle deleted", token_id=created.id)
        except APIError as e:
            logger.error(
                "flow_token_lifecycle delete failed",
                status_code=e.status_code,
                token_id=created.id,
            )


async def main() -> int:
    """Entry point."""
    configure_logging()
    try:
        settings = AuthSettings()  # type: ignore[call-arg]
    except ValidationError as e:
        logger.error("config invalid", errors=e.errors())
        return 2

    await run_api_key(settings)
    login_result = await run_login(settings)

    if login_result is not None:
        user_id, token = login_result
        try:
            await run_token_lifecycle(settings, user_id, token)
        except CanvusError as e:
            logger.error("flow_token_lifecycle unexpected error", error=str(e))
    else:
        logger.warning(
            "flow_token_lifecycle skipped",
            reason="no login session available — set CANVUS_EMAIL/PASSWORD to drive flow 3",
        )
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: go/examples/core/02-auth-flows/main.go
// Command 02-auth-flows demonstrates the three authentication paths the
// Canvus SDK supports:
//
//  1. Static API key (Private-Token header) — long-lived, suitable for daemons.
//  2. Email + password login — exchanges credentials for a short-lived bearer
//     token, suitable for interactive tools.
//  3. Access-token CRUD — creates a programmatic token on behalf of the
//     logged-in user, lists it back, and deletes it.
//
// Each path runs to completion regardless of whether the previous path
// succeeded, so partial credential setups still produce useful output.
package main

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"time"

	"github.com/jaypaulb/MT-Canvus-Tools/go/sdk/canvus"
)

func main() {
	setupLogging()
	if err := run(); err != nil {
		slog.Error("auth-flows failed", "err", err)
		os.Exit(1)
	}
}

func run() error {
	baseURL, err := mustEnv("CANVUS_BASE_URL")
	if err != nil {
		return err
	}
	ctx := context.Background()

	slog.Info("=== auth path 1: API key ===")
	if err := runAPIKey(ctx, baseURL); err != nil {
		slog.Warn("api-key path failed", "err", err)
	}

	slog.Info("=== auth path 2: email + password login ===")
	loginSession, err := runLogin(ctx, baseURL)
	if err != nil {
		slog.Warn("login path failed — skipping access-token lifecycle", "err", err)
		return nil
	}

	slog.Info("=== auth path 3: access-token CRUD ===")
	if err := runTokenLifecycle(ctx, loginSession); err != nil {
		slog.Warn("token lifecycle failed", "err", err)
	}
	return nil
}

// runAPIKey demonstrates the static API-key flow. The Private-Token header is
// injected by an http.RoundTripper installed by WithAPIKey.
func runAPIKey(ctx context.Context, baseURL string) error {
	apiKey := os.Getenv("CANVUS_API_KEY")
	if apiKey == "" {
		return errors.New("CANVUS_API_KEY not set; skipping API-key path")
	}
	cfg := &canvus.SessionConfig{BaseURL: baseURL}
	s := canvus.NewSession(cfg, canvus.WithAPIKey(apiKey))

	canvases, err := s.ListCanvases(ctx, nil)
	if err != nil {
		return fmt.Errorf("ListCanvases via API key: %w", err)
	}
	slog.Info("api-key auth succeeded", "canvas_count", len(canvases))
	return nil
}

// runLogin demonstrates the email + password flow. Per
// VERIFIED-CORRECTIONS.md §4, the server rejects login bodies that include
// any field other than {email, password, remember} — the SDK sends only
// {email, password}.
//
// Returns the authenticated session so callers can demonstrate dependent
// flows (e.g. access-token CRUD).
func runLogin(ctx context.Context, baseURL string) (*canvus.Session, error) {
	email := os.Getenv("CANVUS_EMAIL")
	password := os.Getenv("CANVUS_PASSWORD")
	if email == "" || password == "" {
		return nil, errors.New("CANVUS_EMAIL and/or CANVUS_PASSWORD not set; skipping login path")
	}

	cfg := &canvus.SessionConfig{BaseURL: baseURL}
	// No WithAPIKey: login itself is anonymous.
	s := canvus.NewSession(cfg)
	if err := s.Login(ctx, email, password); err != nil {
		return nil, fmt.Errorf("Login: %w", err)
	}
	userID := s.UserID()
	slog.Info("login succeeded", "user_id", userID)

	// Trivial authenticated call to prove the token works.
	user, err := s.GetUser(ctx, userID)
	if err != nil {
		return s, fmt.Errorf("GetUser(self) post-login: %w", err)
	}
	slog.Info("fetched self", "user_id", user.ID, "email", user.Email, "name", user.Name)
	return s, nil
}

// runTokenLifecycle creates a programmatic access token tied to the
// logged-in user, lists tokens back to confirm it appears, then deletes it
// so we leave the campsite clean.
func runTokenLifecycle(ctx context.Context, s *canvus.Session) error {
	userID := s.UserID()
	if userID == 0 {
		return errors.New("session has no user ID; was Login called?")
	}

	// Create.
	createReq := canvus.CreateAccessTokenRequest{
		Name:        fmt.Sprintf("example-02-auth-flows %s", time.Now().UTC().Format(time.RFC3339)),
		Description: "Temporary token created by mt-canvus-tools example 02-auth-flows.",
	}
	token, err := s.CreateAccessToken(ctx, userID, createReq)
	if err != nil {
		return fmt.Errorf("CreateAccessToken: %w", err)
	}
	slog.Info("access token created",
		"token_id", token.ID,
		"name", token.Name,
		"has_plain_token", token.PlainToken != "")

	// List.
	tokens, err := s.ListAccessTokens(ctx, userID)
	if err != nil {
		return fmt.Errorf("ListAccessTokens: %w", err)
	}
	found := false
	for _, t := range tokens {
		if t.ID == token.ID {
			found = true
			break
		}
	}
	slog.Info("listed access tokens",
		"count", len(tokens),
		"found_new_token", found)

	// Delete.
	if err := s.DeleteAccessToken(ctx, userID, token.ID); err != nil {
		return fmt.Errorf("DeleteAccessToken: %w", err)
	}
	slog.Info("access token deleted", "token_id", token.ID)
	return nil
}

// mustEnv returns the value of name or an error if it is empty/unset.
func mustEnv(name string) (string, error) {
	v := os.Getenv(name)
	if v == "" {
		return "", fmt.Errorf("missing required env var %s", name)
	}
	return v, nil
}

// setupLogging configures slog.Default with a text or JSON handler depending
// on LOG_FORMAT.
func setupLogging() {
	var h slog.Handler
	switch os.Getenv("LOG_FORMAT") {
	case "json":
		h = slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	default:
		h = slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	}
	slog.SetDefault(slog.New(h))
}
